### PR TITLE
Upgrade to hyper/http v1.0

### LINKED ISSUE #1427: [Feature]: Port opentelemetry-http to hyper 1.x
### Related Problems?

https://github.com/hyperium/tonic/issues/1307

### Describe the solution you'd like:

Hyper 1.0 has been released and as such it'd be nice to upgrade opentelemetry-http's `HttpClient` to the new version

### Considered Alternatives

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by aumetra**
Fixes #1427

Blocked by:
- https://github.com/hyperium/tonic/pull/1670
- https://github.com/tokio-rs/axum/pull/2693

## Changes

- Update all hyper and http dependencies to v1.x
- Update reqwest to v0.12
- Remove isahc since it's still stuck on http v0.2

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by aumetra** [comment]
Lint needs to have a feature enabled that's transitively enabled if the full workspace is checked, will do that later.  
The external types check is broken because the nightly used is too old and axum unconditionally enables the diagnostic attribute on nightly without the feature flag because they assume everyone is at least on a nightly that is 1.68

**Review by mladedav — COMMENTED** [review]
I have a few minor points. Feel free to ignore it if you don't care about the features as this should never break anyone unless a downstream crate or cargo make breaking changes and at that point all bets are off.

I think this could also be merged without waiting on `tonic` by adding a dependency on `http@0.2` to `opentelemetry-otlp` and using that for `tonic`-specific errors since that's the only place where these types figure as far as I could find.

File: opentelemetry-zipkin/src/lib.rs
Comment (aumetra) [reply opentelemetry-zipkin/src/lib.rs:87]: Honestly, I don't know anymore. I did it once during the "get it running phase" and forgot about it. I'll have a look

Comment (aumetra) [reply opentelemetry-zipkin/src/lib.rs:87]: Yeah. That was for no reason other than "I wanna see how far the test suite gets in the current state, it's 2AM and I wanna see what else I need to fix".  

Gonna fix this and address the rest of the review once I get the time

File: examples/tracing-http-propagator/src/server.rs
Comment (aumetra) [reply examples/tracing-http-propagator/src/server.rs:49]: Fair, I personally don't find the type signature *too* hard to read though, and we save an allocation (and pointer indirection through dynamic dispatch) without too much of a downside.

I'm open to changing it though 🤷‍♀️

**Comment by aumetra** [comment]
Sidenote: The some tests in this repository seem flakey? Locally I sometimes have to rerun the tests a few times before they pass. Weird.

File: examples/logs-basic/Cargo.toml
Comment (cijothomas) [inline examples/logs-basic/Cargo.toml:15]: `serde_json` was added to support this : https://github.com/open-telemetry/opentelemetry-rust/blob/main/examples/logs-basic/src/main.rs#L11-L13 
Could you keep them to reduce the scope of this PR? 

(I am open to revisiting the `serde_json` dep separately.)

File: Cargo.toml
Comment (cijothomas) [inline Cargo.toml:55]: can you add a comment why is this required.

File: examples/logs-basic/Cargo.toml
Comment (aumetra) [reply examples/logs-basic/Cargo.toml:15]: Sorry, sure. I just wrote `cargo machete` to catch anything that might be dead dependencies after the upgrade, seems like I hit a few things I shouldn't have 🙃

File: Cargo.toml
Comment (aumetra) [reply Cargo.toml:60]: Added

File: examples/tracing-http-propagator/src/server.rs
Comment (mladedav) [reply examples/tracing-http-propagator/src/server.rs:49]: It's not really about the type, more the usage. You have two different methods where each returns a different type and then you have to map the response to wrap it in either left or right and then you even have to do the same thing on the top level to return 404.

Personally, I think that this is just distracting from the purpose of the example which is to show how to extract context from incoming requests. I'd even consider just collecting the bytes in the echo handler and just returning `Response<Full<Bytes>>`. Because performance is not the point at all here. Heck, I'd even vote to remove one of the handlers because it doesn't help the example at all.

Again, feel free to keep what you have and ignore this, it's just opinions.

Comment (aumetra) [reply examples/tracing-http-propagator/src/server.rs:49]: Okay, I really overlooked that this was just an example. I agree, in the case of an example, performance (especially such nit-picky optimizations) aren't really necessary.  

I'll probably just replace it with a boxed body

Comment (aumetra) [reply examples/tracing-http-propagator/src/server.rs]: Well, just replaced it and I'm not entirely sure that that's an improvement.

All in all, the example became *longer* and it now has to showcase the idea that `|err: Infallible| match err {}` resolves to the bottom (never) type, and therefore can be used to coerce it into an `hyper::Error`.

**Comment by TommyCpp** [comment]
> The some tests in this repository seem flakey

Yeah feel free to report those test in https://github.com/open-telemetry/opentelemetry-rust/issues/1559. Most of them are because global setup / env vars and we are working on addressing them as they pop up

File: Cargo.toml
Comment (lalitb) [inline Cargo.toml:30]: We won't be able to continue supporting `isahc` with this upgrade?  Not a concern though, if this is not actively used in the Rust community :)

**Comment by aumetra** [comment]
(missed that one during the last rebase)

**Comment by NOBLES5E** [comment]
I was wondering if this is planned to be merged anytime soon?

**Comment by aumetra** [comment]
> I was wondering if this is planned to be merged anytime soon?

I'm intending to get this over the finish line eventually, issue is just that the tonic PR is still pending

**Comment by lalitb** [comment]
@aumetra The [tonic PR](https://github.com/hyperium/tonic/pull/1670) is merged. do you think we can now remove the patch from this PR, and make it ready to review/merge.

**Comment by aumetra** [comment]
As soon as it's released, sure!

**Comment by aumetra** [comment]
Seems like tonic is close to releasing v0.12 with hyper v1 support.  
After they publish it, I'll rebase and remove the patches

https://github.com/hyperium/tonic/pull/1740

File: opentelemetry-otlp/CHANGELOG.md
Comment (djc) [inline opentelemetry-otlp/CHANGELOG.md:10]: Should probably also mention the tonic 0.12 upgrade here.

File: opentelemetry-zipkin/CHANGELOG.md
Comment (djc) [inline opentelemetry-zipkin/CHANGELOG.md:7]: Not sure this is the best way to explain this change? The material change for the -zipkin API seems to be that it upgraded opentelemetry-http, from which it uses the `HyperClient` trait.

**Review by djc — COMMENTED** [review]
(Drive-by feedback:)

This is looking pretty good, looking forward to getting this released! Have some nits for your consideration.

File: opentelemetry-http/CHANGELOG.md
Comment (lalitb) [reply opentelemetry-http/CHANGELOG.md:6]: @djc Do you think can still keep `isahc` client, as it depends on [http 0.2.1](https://crates.io/crates/isahc/1.7.2/dependencies), while this PR is upgrading to http 1.x

Comment (lalitb) [reply opentelemetry-http/CHANGELOG.md:6]: `isahc` removal should be tagged as breaking change.

**Comment by lalitb** [comment]
@aumetra as https://github.com/hyperium/tonic/pull/1740 is merged, can you resolve the conflicts in this PR (along with patch dependency), so that it can be reviewed and merged ?

File: opentelemetry-http/CHANGELOG.md
Comment (djc) [reply opentelemetry-http/CHANGELOG.md:6]: See https://github.com/sagebind/isahc/issues/435. It hasn't seen a release (or a meaningful commit) in 2 years. I think removing it is fine for now -- if someone asks to have it back it's easy to resurrect.

**Comment by aumetra** [comment]
Sorry, had to set my PC up again after moving :P
Gonna rebase it real quick and then address the review comments.

**Comment by aumetra** [comment]
This `::prost::Enumeration issue` (https://github.com/tokio-rs/prost/issues/1097) is a bit annoying. I silenced it for now because it doesn't seem like there is an actual solution to the problem yet.

File: opentelemetry-appender-log/CHANGELOG.md
Comment (aumetra) [reply opentelemetry-appender-log/CHANGELOG.md:10]: Yeah, probably something that happened during a rebase when I wasn't careful enough..

**Comment by aumetra** [comment]
prost v0.13.0 got yanked so we gotta deal with the broken CI for a little bit

**Comment by djc** [comment]
0.13.1 has been released.

**Comment by aumetra** [comment]
Looking pretty good. Just a question to the code owners if they'd prefer to split the isahc removal into a separate PR. I'm fine with doing either.

**Comment by lalitb** [comment]
> Looking pretty good. Just a question to the code owners if they'd prefer to split the isahc removal into a separate PR. I'm fine with doing either.

Thanks @aumetra . As @djc suggested, good to split as separate PR.

**Comment by lalitb** [comment]
> > Looking pretty good. Just a question to the code owners if they'd prefer to split the isahc removal into a separate PR. I'm fine with doing either.
> 
> Thanks @aumetra . As @djc suggested, good to split as separate PR.

Unless you think it would be difficult to split, as `isahc` client, as it depends on [http 0.2.1](https://crates.io/crates/isahc/1.7.2/dependencies).

**Comment by aumetra** [comment]
I would split it up like this: 

- create PR to delete that client
- get that merged
- rebase on main after merge
- merge this

**Comment by TommyCpp** [comment]
+1 on split the client clean up to another PR to control the scope

**Comment by aumetra** [comment]
Removed `isahc` support in <https://github.com/open-telemetry/opentelemetry-rust/pull/1924>, gonna rebase this on that branch so we can move ahead.

File: opentelemetry-zipkin/CHANGELOG.md
Comment (aumetra) [reply opentelemetry-zipkin/CHANGELOG.md]: Yeah, I was absent-mindedly adding that line to changelogs where I changed something re http/hyper v1.  
This is more of an "update examples to hyper v1" situation. Adjusted.

**Comment by aumetra** [comment]
Just rebased it real quick to make the eventual rebase on `main` based on <https://github.com/open-telemetry/opentelemetry-rust/pull/1924> easier.
Waiting for that PR to merge for now.

**Comment by lalitb** [comment]
@aumetra Is the upgrade to `opentelemetry-proto` submodule intended/required? If not, can we remove it from this PR, to be handled separately.

File: opentelemetry-sdk/src/metrics/periodic_reader.rs
Comment (lalitb) [inline opentelemetry-sdk/src/metrics/periodic_reader.rs:294]: curious - Is this change related to hyper upgrade?

File: opentelemetry-zipkin/CHANGELOG.md
Comment (lalitb) [inline opentelemetry-zipkin/CHANGELOG.md:7]: nit - this doesn't look to be user impacting update, can be removed from changelog ?

File: opentelemetry-sdk/CHANGELOG.md
Comment (lalitb) [inline opentelemetry-sdk/CHANGELOG.md:77]: Does hyper upgrade impacts opentelemetry-sdk in anyway?

File: Cargo.toml
Comment (utpilla) [inline Cargo.toml:28]: nit: Re-arrange these to an alphabetically sorted order.

**Comment by aumetra** [comment]
> @aumetra Is the upgrade to `opentelemetry-proto` submodule intended/required? If not, can we remove it from this PR, to be handled separately.

Not intended. Probably mistyped somewhere when attempting to reset the git submodule. Gonna revert that

File: opentelemetry-sdk/src/metrics/periodic_reader.rs
Comment (djc) [reply opentelemetry-sdk/src/metrics/periodic_reader.rs:294]: They're equivalent, but I do prefer the new version -- better to keep "boring" traits at the end.

File: opentelemetry-zipkin/CHANGELOG.md
Comment (djc) [reply opentelemetry-zipkin/CHANGELOG.md:7]: Referring to the examples here is a little weird, but opentelemetry-zipkin does expose the `HyperClient` trait in its `ZipkinPipelineBuilder` API so IMO there should be some mention of this change.

File: opentelemetry-sdk/CHANGELOG.md
Comment (djc) [reply opentelemetry-sdk/CHANGELOG.md:77]: The `HttpClient` trait appears in the `Sampler::jaeger_remote()` API.

File: opentelemetry-sdk/src/metrics/periodic_reader.rs
Comment (aumetra) [reply opentelemetry-sdk/src/metrics/periodic_reader.rs:294]: Yeah, that was part of the intention here. Also I had some linter issues with clippy on nightly lints a while back, I think?
Can't remember, but this is probably an artifact of that.

File: opentelemetry-zipkin/CHANGELOG.md
Comment (aumetra) [reply opentelemetry-zipkin/CHANGELOG.md:7]: Does it though? I haven't found the point where it's exposed that you're referring to

Comment (aumetra) [reply opentelemetry-zipkin/CHANGELOG.md:7]: Oh, do you mean `HttpClient`? Fair. Yeah.

Comment (djc) [reply opentelemetry-zipkin/CHANGELOG.md:7]: Sorry, yes, `HttpClient`.

Comment (aumetra) [reply opentelemetry-zipkin/CHANGELOG.md]: Adjusted the changelog entry. Let me know what you think.

Comment (djc) [reply opentelemetry-zipkin/CHANGELOG.md]: LGTM!

File: opentelemetry-http/src/lib.rs
Comment (djc) [inline opentelemetry-http/src/lib.rs:117]: BTW, have you considered abstracting over `B` here? Like, picking some concrete type (maybe vendoring `http_body_util::Full`) instead of exposing this as a type parameter on `HypercClient`?

Comment (aumetra) [reply opentelemetry-http/src/lib.rs]: What about this solution I just pushed?
I like the idea of this wrapper more than vendoring anything.

Comment (djc) [reply opentelemetry-http/src/lib.rs]: Yup, that's probably better in this case! One nit: that file is laid out in top-down order, so would be better if the `Body` wrapper moves down to somewhere near the bottom of the module (probably before `ResponseExt`).

**Review by djc — APPROVED** [review]

File: opentelemetry-http/src/lib.rs
Comment (aumetra) [reply opentelemetry-http/src/lib.rs]: Even though we leak some information about the inner impl that could *technically* cause some headaches if it's ever swapped out for something else.
The `Infallible` type. Maybe this should become a box dyn error?

Comment (djc) [reply opentelemetry-http/src/lib.rs]: Yeah, that seems reasonable.

File: opentelemetry-sdk/CHANGELOG.md
Comment (aumetra) [reply opentelemetry-sdk/CHANGELOG.md]: Updated the changelog note to be a bit more detailed

File: opentelemetry-proto/src/proto/tonic/opentelemetry.proto.collector.logs.v1.rs
Comment (lalitb) [inline opentelemetry-proto/src/proto/tonic/opentelemetry.proto.collector.logs.v1.rs:338]: @aumetra  - Just to be sure, are these files regenerated by running [grpc_build.rs](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-proto/tests/grpc_build.rs) against the new tonic version, and not manually modified ?

Comment (aumetra) [reply opentelemetry-proto/src/proto/tonic/opentelemetry.proto.collector.logs.v1.rs:338]: Yeah, I ran the `cargo test` command to generate it using the tests

**Comment by djc** [comment]
@lalitb are you planning a release? Need help with that?

**Comment by lalitb** [comment]
> @lalitb are you planning a release? Need help with that?

@djc yes I believe @cijothomas plans to do a release today if there are no issues.

**Comment by djc** [comment]
Great, looking forward to it!

**Pull request merged**
